tests/unit/steps/test_steps_execute.py

```diff
@@ -16,85 +16,215 @@
+        PreUpgradeStep("pre-upgrade"),
+        UpgradeStep("upgrade 1"),
+        UpgradeStep("upgrade 2 halt"),
+        UpgradeStep("upgrade 3 halt"),
+        UpgradeStep("upgrade 4"),
+        PostUpgradeStep("post-upgrade"),
+    ]
+    mock_apply_step.side_effect = _apply_step
+
+    with pytest.raises(HaltUpgradeExecution):
+        await _run_sub_stpes_in_parallel(upgrade_step, False, False)
+
+    assert finished_steps == ["pre-upgrade", "upgrade 1", "upgrade 4", "post-upgrade"]
+    assert failed_steps == ["upgrade 2 halt", "upgrade 3 halt"]
+    mock_apply_step.assert_has_awaits([call(step, False, False) for step in sub_steps])
 
 
 @pytest.mark.asyncio
 @patch("cou.steps.execute.apply_step")
-async def test_run_step_in_parallel(mock_apply_step):
-    """Test running step and all sub-steps in parallel."""
-    upgrade_step = AsyncMock(spec_set=UpgradeStep())
+async def test_run_sub_stpes_sequentially(mock_apply_step):
+    """Test running all sub-steps of step sequentially."""
+    upgrade_step = MagicMock(spec_set=UpgradeStep())
+    upgrade_step.parallel = True
+    upgrade_step.sub_steps = sub_steps = [
+        PreUpgradeStep("pre-upgrade"),
+        UpgradeStep("upgrade"),
+        PostUpgradeStep("post-upgrade"),
+    ]
+
+    await _run_sub_stpes_sequentially(upgrade_step, False, False)
+    mock_apply_step.assert_has_awaits([call(step, False, False) for step in sub_steps])
+
+
+@pytest.mark.asyncio
+@patch("cou.steps.execute.apply_step")
+async def test_run_sub_stpes_sequentially_halt(mock_apply_step):
+    """Test running all sub-steps of step sequentially."""
+    upgrade_step = MagicMock(spec_set=UpgradeStep())
     upgrade_step.parallel = True
     upgrade_step.sub_steps = sub_steps = [
-        AsyncMock(spec_set=PreUpgradeStep()),
-        AsyncMock(spec_set=UpgradeStep()),
-        AsyncMock(spec_set=PostUpgradeStep()),
+        PreUpgradeStep("pre-upgrade"),
+        UpgradeStep("upgrade"),
+        PostUpgradeStep("post-upgrade"),
     ]
+    mock_apply_step.side_effect = [None, HaltUpgradeExecution("halt"), None]
+
+    with pytest.raises(HaltUpgradeExecution):
+        await _run_sub_stpes_sequentially(upgrade_step, False, False)
+
+    # Note(rgildein): Since the second step is raising HaltUpgradeExecution, apply_plan will be
+    #                 awaited only 2 times.
+    mock_apply_step.assert_has_awaits(
+        [call(sub_steps[0], False, False), call(sub_steps[1], False, False)]
+    )
+
+
+@pytest.mark.asyncio
+@patch("cou.steps.execute.apply_step")
+@patch("cou.steps.execute._run_sub_stpes_sequentially")
+@patch("cou.steps.execute.progress_indicator")
+async def test_run_step_sequentially_upgrade_step(
+    mock_indicator, mock_run_sub_stpes_sequentially, mock_apply_step
+):
+    """Test running upgrade step and all sub-steps sequentially."""
+    upgrade_step = MagicMock(spec_set=UpgradeStep())
+    upgrade_step.run = AsyncMock()
+    upgrade_step.parallel = False
```

Comment:
wouldn't make sense to add sub_steps here to match the description?

This and the next tests says about sub_steps but they don't have it.

Reply:
There is line testing that `mock_run_sub_steps_sequentially.assert_awaited_once_with(upgrade_step, False, False)`, and since we are not touching sub_steps in this test, it does not make sense add any.